process/block/gasConsumption.go

```diff
@@ -0,0 +1,173 @@
+package block
+
+import (
+	"sync"
+
+	"github.com/multiversx/mx-chain-core-go/core/check"
+	"github.com/multiversx/mx-chain-core-go/data"
+	"github.com/multiversx/mx-chain-go/process"
+	"github.com/multiversx/mx-chain-go/sharding"
+)
+
+const (
+	initialLimitsFactor = uint64(100)
+	// TODO: move these to config
+	percentSplitBlock         = uint64(50) // 50%
+	percentDecreaseLimitsStep = uint64(10) // 10%
+	minPercentLimitsFactor    = uint64(10) // 10%
+)
+
+type gasConsumption struct {
+	sync.RWMutex
+	economicsFee            process.FeeHandler
+	shardCoordinator        sharding.Coordinator
+	totalGasConsumedInShard map[uint32]uint64
+	gasConsumedByMiniBlock  map[string]uint64
+	limitsFactor            uint64
+}
+
+// NewGasConsumption returns a new instance of gasConsumption
+func NewGasConsumption(
+	economicsFee process.FeeHandler,
+	shardCoordinator sharding.Coordinator,
+) (*gasConsumption, error) {
+	if check.IfNil(economicsFee) {
+		return nil, process.ErrNilEconomicsData
+	}
+	if check.IfNil(shardCoordinator) {
+		return nil, process.ErrNilShardCoordinator
+	}
+
+	return &gasConsumption{
+		economicsFee:            economicsFee,
+		shardCoordinator:        shardCoordinator,
+		totalGasConsumedInShard: make(map[uint32]uint64),
+		gasConsumedByMiniBlock:  make(map[string]uint64),
+		limitsFactor:            initialLimitsFactor,
+	}, nil
+}
+
+// CheckTransactionsForMiniBlock checks if the provided transactions can be added in a mini block
+// for the current selection session
+func (gc *gasConsumption) CheckTransactionsForMiniBlock(
+	miniBlockHash []byte,
```

Comment:
for the proposed intrashard miniblock, we don't have yet a mbHash, we need to first see which transactions can fit into the current limits.